### PR TITLE
tools: create CephFS shell

### PR DESCRIPTION
**Pull request by rpavani1998**
a shell and CLI tool interface to CephFS using the existing libcephfs client library and python bindings. Implemented few commands like mkdir, put, get and so on.
[http://tracker.ceph.com/issues/24286](http://tracker.ceph.com/issues/24286)
Signed-off-by: Pavani Rajula <rpavani1998@gmail.com>

### CODE REVIEW

File: src/pybind/cephfs/cephfs.pyx [hunk 1]
```diff
@@ -143,8 +143,9 @@ cdef extern from "cephfs/libcephfs.h" nogil:
     int ceph_sync_fs(ceph_mount_info *cmount)
     int ceph_fsync(ceph_mount_info *cmount, int fd, int syncdataonly)
     int ceph_conf_parse_argv(ceph_mount_info *cmount, int argc, const char **argv)
+    int ceph_chmod(ceph_mount_info *cmount, const char *path, mode_t mode)
```
Comment (tchaikov) [inline src/pybind/cephfs/cephfs.pyx]: you `chmod`ed cephfs.pyx to 755. could you revert this change?

File: src/tools/cephfs/cephfs_shell.py [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+    dir_ = cephfs.getcwd()
+    is_rel_path = True
+    if pattern[0] == '/':
+        dir_ = '/'
+        pattern = pattern[1:]
+        is_rel_path = False
+    patterns = pattern.split('/')
+    paths.extend(glob(dir_, patterns[0]))
+    patterns.pop(0) 
+    for index in range(len(patterns)) :
```
Comment (tchaikov) [inline src/tools/cephfs/cephfs_shell.py]: nit, why not put
```py
for pattern in patterns:
```
?

File: src/tools/cephfs/cephfs_shell.py [hunk 3]
```diff
@@ -0,0 +1,848 @@
...
+    def do_rmdir(self, args):
+        """ 
+        Remove a specific Directory
+            Usage : rmdir <dir_path>
+        """
+        try:
+            is_pattern = False
+            directories = args.dir_paths
+            for i, dir_path in enumerate(directories):
+                if dir_path.count('*') > 0 :
```
Comment (tchaikov) [inline src/tools/cephfs/cephfs_shell.py]: remove space before `:`.

File: src/tools/cephfs/cephfs_shell.py [hunk 4]
```diff
@@ -0,0 +1,848 @@
...
+    @with_argparser(rmdir_parser)
+    def do_rmdir(self, args):
+        """ 
+        Remove a specific Directory
+            Usage : rmdir <dir_path>
+        """
+        try:
+            is_pattern = False
+            directories = args.dir_paths
+            for i, dir_path in enumerate(directories):
```
Comment (tchaikov) [inline src/tools/cephfs/cephfs_shell.py]: nit, the index is not used.

File: src/tools/cephfs/cephfs_shell.py [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+
+    @with_argparser(chmod_parser)
+    def do_chmod(self, args):
+        """
+        Change permission of a file
+        """
+        try:           
+            cephfs.chmod(args.file_name, args.mode)
+        except :
+             traceback.print_exc(file = sys.stdout)
```
Comment (tchaikov) [inline src/tools/cephfs/cephfs_shell.py]: this pattern is repeating itself in this script. could make it a decorator also.

File: src/tools/cephfs/cephfs_shell.py [hunk 6]
```diff
@@ -0,0 +1,848 @@
...
+        """
+        Lists all files and folders in the current local directory
+        """
+        try:
+            if len(args.paths) == 1 and args.paths[0] == '':
+                args.paths.pop(0)
+                args.paths.append(os.getcwd())
+            for path in args.paths :
+                print(path)
+                if not path.startswith(os.getcwd()) and not path.startswith('/'):
```
Comment (tchaikov) [inline src/tools/cephfs/cephfs_shell.py]: could use `os.path.isabs()` instead.

File: src/tools/cephfs/cephfs_shell.py [hunk 7]
```diff
@@ -0,0 +1,848 @@
...
+                    path = os.getcwd() + '/'+ path
```
Comment (tchaikov) [inline src/tools/cephfs/cephfs_shell.py]: might want to use `os.path.relpath()`

File: doc/cephfs/shell.rst [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+alias
+-----
+
+ Define or display aliases
+
+ Usage: alias [name] | [<name> <value>]
+     Where:
+         name - name of the alias being looked up, added, or replaced
+         value - what the alias will be resolved to (if adding or replacing)
+                 this can contain spaces and does not need to be quoted
```
Comment (joscollin) [inline doc/cephfs/shell.rst]: Build fails here. `shell.rst:174:Unexpected indentation.`

File: doc/cephfs/shell.rst [hunk 9]
```diff
@@ -0,0 +1,277 @@
+
+Ceph FS Shell
+=============
+
+The File System (FS) shell includes various shell-like commands that directly interact with the Ceph File System.
```
Comment (joscollin) [inline doc/cephfs/shell.rst]: See: https://github.com/ceph/ceph/pull/22784#pullrequestreview-135138797 and follow it everywhere else.

Comment (rpavani1998) [reply doc/cephfs/shell.rst]: Thanks! I will work on it.

File: src/pybind/cephfs/cephfs.pyx (hunk 1 above)
Comment (rpavani1998) [reply src/pybind/cephfs/cephfs.pyx]: you mean the cephfs.pyx file permission right? I will do it.

File: doc/cephfs/shell.rst (hunk 9 above)
Comment (joscollin) [reply doc/cephfs/shell.rst]: 1. Put the doc changes in a separate commit.
2. The commit lack `Fixes:` tag. See [Tag the Commit](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#tag-the-commit)
3. Update the tracker with the PR url and change the status to Needs Review.

File: src/pybind/cephfs/cephfs.pyx (hunk 1 above)
Comment (tchaikov) [reply src/pybind/cephfs/cephfs.pyx]: yup, the permission bits.

File: doc/cephfs/shell.rst [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+Commands
+========
+
+mkdir
+-----
+ Create the directory(ies), if they do not already exist.
+
+ Usage : mkdir [-option] <directory>...
```
Comment (joscollin) [inline doc/cephfs/shell.rst]: Put the `Usage` in a box? See: https://github.com/ceph/ceph/commit/8b796173063ac9af8c21364521fc5ee23d901196

File: doc/cephfs/shell.rst [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+
+ Usage : history [-h] [-r | -e | -s | -o FILE | -t TRANSCRIPT] [arg]
+
+ Options:
+   -h            show this help message and exit
+   -r            run selected history items
+   -e            edit and then run selected history items
+   -s            script format; no separation lines
+   -o FILE       output commands to a script file
+   -t TRANSCRIPT output commands and results to a transcript file
```
Comment (joscollin) [inline doc/cephfs/shell.rst]: box?

File: doc/cephfs/shell.rst [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+unalias
+-------
+
+ Unsets aliases
+ 
+ Usage : unalias [-a] name [name ...]
+     Where:
+         name - name of the alias being unset
```
Comment (joscollin) [inline doc/cephfs/shell.rst]: See: http://docs.ceph.com/docs/master/man/8/ceph/. This style eliminates the repeated `Where:`

For example: 
```
Subcommand remove removes monitor named <name>.
Usage:
   ceph mon remove <name>
```

File: src/tools/cephfs/cephfs_shell.py [hunk 13]
```diff
@@ -0,0 +1,848 @@
...
+import fnmatch
+import math
+cephfs = None
+
+def setup_cephfs():
+    """ 
+    Mouting a cephfs 
+    """
+    global cephfs
+    cephfs = libcephfs.LibCephFS(conffile = '/home/admin1/Documents/ceph/build/ceph.conf')
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `libcephfs.LibCephFS(conffile = '')`

Should be ^

File: doc/cephfs/shell.rst [hunk 14]
```diff
@@ -0,0 +1,277 @@
+
```
Comment (batrick) [inline doc/cephfs/shell.rst]: Please add:

`.. _cephfs-shell:`

at the top.

We will use that to reference to this document in the index.

File: src/tools/cephfs/cephfs_shell.py [hunk 15]
```diff
@@ -0,0 +1,848 @@
...
+            traceback.print_exc(file = sys.stdout)
+   
+    def complete_rmdir(self, text, line, begidx, endidx):
+        """ 
+        auto complete of file name.
+        """    
+        return self.complete_filenames(text, line, begidx, endidx)
+
+    rmdir_parser = argparse.ArgumentParser(description = 'Remove Directory.')
+    rmdir_parser.add_argument('dir_paths', help = 'Directory Path.',nargs = '*', default = [''])
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `nargs='+'`. No default.

File: src/tools/cephfs/cephfs_shell.py [hunk 16]
```diff
@@ -0,0 +1,848 @@
...
+        auto complete of file name.
+        """    
+        return self.complete_filenames(text, line, begidx, endidx)
+        
+    ls_parser = argparse.ArgumentParser(description = 'Copy a file from Ceph File System from Local Directory.')
+    ls_parser.add_argument('-l', '--long', action = 'store_true', help = 'Detailed list of items in the directory.')
+    ls_parser.add_argument('-r', '--reverse', action = 'store_true', help = 'Reverse order of listing items in the directory.')
+    ls_parser.add_argument('-H', action = 'store_true', help = 'Human Readable')
+    ls_parser.add_argument('-S', action = 'store_true', help = 'Sort by file_size')
+    ls_parser.add_argument('dir_names', help = 'Name of Directories', nargs = '*', default = [''])
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `nargs='*', default=['.']`

Should simplify the processing below.

File: src/tools/cephfs/cephfs_shell.py [hunk 17]
```diff
@@ -0,0 +1,848 @@
...
+    
+    def complete_get(self, text, line, begidx, endidx):
+        """ 
+        auto complete of file name.
+        """    
+        return self.complete_filenames(text, line, begidx, endidx)
+
+    get_parser = argparse.ArgumentParser(description = 'Copy a file from Ceph File System from Local Directory.')
+    get_parser.add_argument('remote_path', type = str, help = 'Path of the file in the remote system')
+    get_parser.add_argument('local_path', type = str, help = 'Path of the file in the local system')
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `local_path` should be optional. The default should be `os.path.basename(remote_path)`.

File: src/tools/cephfs/cephfs_shell.py (hunk 3 above)
Comment (batrick) [reply src/tools/cephfs/cephfs_shell.py]: (@rpavani1998: To be clear, @tchaikov is asking you to follow the existing Python whitespace rules used in the rest of Ceph.)

File: src/tools/cephfs/cephfs_shell.py [hunk 18]
```diff
@@ -0,0 +1,848 @@
...
+            traceback.print_exc(file = sys.stdout)
+   
+    def complete_cd(self, text, line, begidx, endidx):
+        """ 
+        auto complete of file name.
+        """    
+        return self.complete_filenames(text, line, begidx, endidx)
+
+    cd_parser = argparse.ArgumentParser(description = 'Create Directory.')
+    cd_parser.add_argument('dir_name', type = str, help = 'Name of the directory.', nargs = '*', default = [''])
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `default='.'`

and remove `nargs`

File: src/tools/cephfs/cephfs_shell.py [hunk 19]
```diff
@@ -0,0 +1,848 @@
...
+
+    @with_argparser(cd_parser)
+    def do_cd(self, args):
+        """ 
+        Open a specific directory.
+            Usage : cd <dir_name>
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `cd [dir_name]`

Closed brackets indicate the argument is optional. Make sure the rest of the documentation is consistent.

File: src/tools/cephfs/cephfs_shell.py [hunk 20]
```diff
@@ -0,0 +1,848 @@
...
+        return self.complete_filenames(text, line, begidx, endidx)
+
+    chmod_parser = argparse.ArgumentParser(description = 'Create Directory.')
+    chmod_parser.add_argument('file_name', type = str, help = 'Name of the file')
+    chmod_parser.add_argument('mode', type = int, help = 'Mode')
+
+    @with_argparser(chmod_parser)
+    def do_chmod(self, args):
+        """
+        Change permission of a file
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `chmod <mode> <path>`

Please reverse the arguments.

File: src/tools/cephfs/cephfs_shell.py [hunk 21]
```diff
@@ -0,0 +1,848 @@
...
+        """
+        try:
+            for file_name in args.file_names :
+                print(file_name)
+                copy_to_local(file_name, '-')
+        except :
+            traceback.print_exc(file = sys.stdout)
+   
+    umask_parser = argparse.ArgumentParser(description = '')
+    umask_parser.add_argument('mode', help = 'Mode', action  = 'store', nargs = '*', default = [''])
```
Comment (batrick) [inline src/tools/cephfs/cephfs_shell.py]: `umask_parser.add_argument('mode', help = 'Mode', type=lambda x: return int(x, 8))`

(untested)

**Comment by batrick** [comment]
@rpavani1998 please also rebase this on the latest master.

File: doc/cephfs/shell.rst (hunk 14 above)
Comment (joscollin) [reply doc/cephfs/shell.rst]: Then update `cephfs/index.rst` too in this PR.

**Comment by batrick** [comment]
@rpavani1998 please rebase again, your umask PR is merged into master.

File: src/tools/cephfs/cephfs-shell [hunk 22]
```diff
@@ -0,0 +1,830 @@
...
+            return
+        parser = self.create_argparser(line)
+        if parser:
+            parser.print_help()
+        else:
+            super().do_help(line)
+
+ 
+if __name__ == '__main__':
+    setup_cephfs()
```
Comment (batrick) [inline src/tools/cephfs/cephfs-shell]: Add argument parsing here. At the very least we'll want a `-c` option which lets' us set the config file here: https://github.com/ceph/ceph/pull/23158/files#diff-9226892dd49173c4687a9953904ccaa8R24

File: src/tools/cephfs/cephfs-shell [hunk 23]
```diff
@@ -0,0 +1,830 @@
...
+            else:
+                for x in dirwalk(fullpath): 
+                    if giveDirs:
+                        yield x 
+                    else:
+                        yield x 
+        else:
+            yield re.sub('\/+', '/',fullpath)
+
+class CephfsShell(Cmd):
```
Comment (batrick) [inline src/tools/cephfs/cephfs-shell]: `CephFSShell` please

File: src/tools/cephfs/cephfs-shell [hunk 24]
```diff
@@ -0,0 +1,830 @@
...
+        """
+        cephfs.rename(to_bytes(args.src_path), to_bytes(args.dest_path))
+
+    def complete_cd(self, text, line, begidx, endidx):
+        """ 
+        auto complete of file name.
+        """    
+        return self.complete_filenames(text, line, begidx, endidx)
+
+    cd_parser = argparse.ArgumentParser(description = 'Create Directory.')
```
Comment (batrick) [inline src/tools/cephfs/cephfs-shell]: `'Change working directory'`

**Comment by batrick** [comment]
@rpavani1998 looks like you removed the shell doc? in feb96f7cb4b6c6a849bd15499eaa74d5c777cbd3

**Comment by rpavani1998** [comment]
I have changed the doc path to  doc/man/8/cephfs-shell.rst

**Comment by batrick** [comment]
@rpavani1998, Kefu's man page is a good idea but it requires reformatting and existing templates like [1] don't render well in a browser. Let's keep your cephfs-shell doc in doc/cephfs/ for now.

[1] https://github.com/ceph/ceph/blob/ce45875d80150eb8ee6bacdff14b1de3a2e62161/doc/man/8/ceph.rst

**Comment by rpavani1998** [comment]
@batrick updated.

**Comment by batrick** [comment]
@rpavani1998 FYI, I amended your commit message to fit style-guidelines and explain what the shell is.

File: src/pybind/cephfs/cephfs.pyx (hunk 1 above)
Comment (tchaikov) [reply src/pybind/cephfs/cephfs.pyx]: @rpavani1998 #23417

**Comment by rpavani1998** [comment]
@batrick Okay.Thanks!
